# Task 002 - Data Preparation

## CRISP-DM Phase 3

---

# مقدمه

در تسک قبلی شما مرحله **Data Understanding** را روی دیتاست بیماری قلبی انجام دادید.

در پایان آن مرحله باید به درک مناسبی از موارد زیر رسیده باشید:

* ساختار داده
* کیفیت داده
* ویژگی‌های مهم
* ارتباط ویژگی‌ها با متغیر هدف
* مشکلات موجود در داده
* مناسب بودن داده برای مدل‌سازی

اکنون وارد مرحله سوم فرآیند CRISP-DM یعنی **Data Preparation** می‌شویم.

---

# یادآوری چالش کسب‌وکار

شما به عنوان تحلیلگر داده در یک مرکز درمانی قلب و عروق فعالیت می‌کنید.

هدف کسب‌وکار:

تشخیص احتمال ابتلا به بیماری قلبی بر اساس ویژگی‌های بیماران.

متغیر هدف:

```python
target
```

* 0 → Healthy
* 1 → Heart Disease

---

# نوع مسئله داده‌کاوی

با توجه به تحلیل مرحله قبل:

متغیر هدف مشخص است و دو کلاس دارد.

بنابراین مسئله اصلی:

## Supervised Learning

و به طور دقیق‌تر:

## Binary Classification

---

# الگوریتم‌های احتمالی آینده

در مراحل بعدی ممکن است از الگوریتم‌های زیر استفاده شود:

* Logistic Regression
* Decision Tree
* Random Forest
* KNN
* SVM
* Naive Bayes

بنابراین باید داده را به شکلی آماده کنیم که برای این الگوریتم‌ها مناسب باشد.

---

# هدف این تسک

ایجاد یک دیتاست نهایی و آماده برای مرحله Modeling

---

# خروجی نهایی مورد انتظار

در پایان این تسک باید فایل زیر را ایجاد کنید:

```python
heart_prepared.csv
```

این فایل باید:

* تمیز باشد
* ویژگی‌های انتخاب شده را داشته باشد
* ویژگی‌های جدید مورد نیاز را داشته باشد
* فرمت مناسبی داشته باشد
* داده‌های عددی مقیاس‌بندی شده باشند
* برای مدل‌سازی آماده باشد

---

# قوانین

- استفاده از Scikit-Learn ممنوع
- استفاده از AutoML ممنوع


فقط از موارد زیر استفاده شود

- NumPy
- Pandas
- Matplotlib
- Seaborn

---

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("../task_1/heart.csv")
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1



ویژگی|ترجمه|نوع
|---|---|---|
|age | سن|عددی
|sex |جنسیت|دسته ای
|cp |نوع درد قفسه سینه |دسته ای
|trestbps |فشار خون استراحت|عددی
|chol |کلسترول|عددی
|fbs | قند خون ناشتا|دودویی
|restecg |نوار قلب|دسته ای
|thalach |حداکثر ضربان قلب|عددی
|exang |آنژین ناشی از ورزش|دودویی
|oldpeak |stافت موج |عددی
|slope |شیب قطعه اس تی در ورزش|دسته ای
|ca |تعداد عروق اصلی|دسته ای
|thal |تالاسمی|دسته ای
|target |متغییر هدف |دودویی 

# بخش اول

# Data Selection

---

## سوال 1

براساس تحلیل مرحله قبل:

کدام ستون‌ها بیشترین ارتباط را با target داشتند؟

فهرست آن‌ها را استخراج کنید.

---

In [23]:
corr_with_target = df.corr(numeric_only=True)["target"].sort_values(ascending=False)
print(corr_with_target)

top_features = corr_with_target.drop("target").abs().sort_values(ascending=False)
print("\nmost important features (absolute values):")
print(top_features)

target      1.000000
cp          0.433798
thalach     0.421741
slope       0.345877
restecg     0.137230
fbs        -0.028046
chol       -0.085239
trestbps   -0.144931
age        -0.225439
sex        -0.280937
thal       -0.344029
ca         -0.391724
oldpeak    -0.430696
exang      -0.436757
Name: target, dtype: float64

most important features (absolute values):
exang       0.436757
cp          0.433798
oldpeak     0.430696
thalach     0.421741
ca          0.391724
slope       0.345877
thal        0.344029
sex         0.280937
age         0.225439
trestbps    0.144931
restecg     0.137230
chol        0.085239
fbs         0.028046
Name: target, dtype: float64


ghavi tarin ravabet: exang, cp & oldpeak.
chon faghat be meghdare hambastegi niaz darim, absolute migirim. 

## سوال 2

یک جدول تهیه کنید:

---



| Feature | Business Importance | Data Importance(corr) | analysis |
| ------- | ------------------- | --------------- | -------- |
| cp | Very high — Chest pain type is one of the most direct clinical indicators |	High (0.434) |Consistent — Clinical and statistical importance agree|
| thalach | High — Maximum heart rate achieved during exercise | High (0.422) | Consistent |
| oldpeak | High — ST depression on ECG, a strong diagnostic indicator | High (0.431) | Consistent |
| exang	| High — Exercise-induced angina | High (0.437) | Consistent |
| ca | High — Number of major vessels colored by fluoroscopy | High (0.392)	| Consistent |
| thal | High — Thallium stress test result	| Moderate (0.344) | Fairly consistent |
| chol | High — Cholesterol is a classic medical risk factor | Low (0.085) | Inconsistent — Clinically important, but shows weak correlation in this dataset |
| trestbps | High — Resting blood pressure | Low (0.145) | Partially inconsistent |
| fbs | Moderate — Fasting blood sugar | Very low (0.028) | Inconsistent — Medically important, but has little predictive value in this dataset | 
| age | High — Age is an important risk factor | Moderate (0.225) | Fairly consistent |


*included research*

## سوال 3

حداقل سه ویژگی را انتخاب کنید که احتمال می‌دهید برای مدل‌سازی مفید نباشند.

دلایل خود را بنویسید.

---

In [9]:
for col in ["fbs", "chol", "restecg"]:
    print(col, "-> correlation with target:", abs(df[col].corr(df["target"])))

fbs -> correlation with target: 0.028045760272712838
chol -> correlation with target: 0.08523910513756908
restecg -> correlation with target: 0.13722950287377303


fbs: 0.028 correlation ba target. kamtarin meghdar dar kole dataset. in sotun taghriban hich etelaate tamayoz dahandeii beine afrade salem va bimar eraee nemide. hamchenin be dalile mahiate binary va skewness shadid, toziie besiar asymetric y dare.

chol: 0.085 correlation ba target. ba vojood inke az lahaze pezeshki amele mohemie, vali tooye in dataset rabete zaiifi dare ba target. toziie in sotun ham daraye skewness ziad va outlier ziade. 

restECG: 0.137 correlation ba target. in sotun faghat se meghdare 0, 1 va 2 dare va toziiesh besiar asymmetric e. aksare bimara meghdare 0 ya 1 daran. pas ghodrate tamayoz mahdoodi dare. 


## سوال 4

یک نسخه جدید از دیتاست بسازید که فقط شامل ویژگی‌های منتخب شما باشد.

نام:

```python
selected_df
```

---

In [10]:
selected_columns = ["age", "sex", "cp", "trestbps", "thalach",
                    "exang", "oldpeak", "ca", "thal", "target"]

selected_df = df[selected_columns].copy()
selected_df.head()

,age,sex,cp,trestbps,thalach,exang,oldpeak,ca,thal,target
0,63,1,3,145,150,0,2.3,0,1,1
1,37,1,2,130,187,0,3.5,0,2,1
2,41,0,1,130,172,0,1.4,0,2,1
3,56,1,1,120,178,0,0.8,0,2,1
4,57,0,0,120,163,1,0.6,0,2,1


ba tavajoh be soalate ghabl, sotun haye chol, fbs، restecg va slope az dataset hazf shodan. 
(slope taghriban mesle oldpeake va correlation balaii daran)


## سوال 5

تعداد ویژگی‌ها قبل و بعد از Selection را مقایسه کنید.

---

In [26]:
print("تعداد ستون‌ها قبل : ", df.shape[1])
print("تعداد ستون‌ها بعد : ", selected_df.shape[1])
print("تعداد ستون‌های حذف‌شده: ", df.shape[1] - selected_df.shape[1])
print("درصد ستون های حذف شده: ", (df.shape[1] - selected_df.shape[1]) / df.shape[1] * 100)

تعداد ستون‌ها قبل :  14
تعداد ستون‌ها بعد :  10
تعداد ستون‌های حذف‌شده:  4
درصد ستون های حذف شده:  28.57142857142857


# بخش دوم

# Data Cleaning

---

## سوال 6

تمام مقادیر Null را مجدداً بررسی کنید.

---

In [37]:
print(selected_df.isnull().sum())
print("\nsum of null values:", selected_df.isnull().sum().sum())

age         0
sex         0
cp          0
trestbps    0
thalach     0
exang       0
oldpeak     0
ca          0
thal        0
target      0
dtype: int64

sum of null values: 0


## سوال 7

برای هر ستون مشخص کنید:

* تعداد Null
* درصد Null

---

In [29]:
# null_count = selected_df.isnull().sum()
# null_percent = (selected_df.isnull().sum() / len(selected_df)) * 100

# null_report = pd.DataFrame({
#     "Null Count": null_count,
#     "Null Percent": null_percent.round(2)
# })
# null_report = null_report[null_report["Null Count"] > 0].sort_values("Null Count", ascending=False)
# print(null_report)

## سوال 8

اگر داده گمشده وجود دارد:

سه روش مختلف برای پر کردن آن پیشنهاد دهید.

---

Mean/Median Imputation: baraye sotun hate numerical continious mesle trestbps & thalach, Median behtar az mean ast chon in sotun ha Outlier daran va agar mean begirim be samte outlier keshide mishe. 

Mode Imputation: baraye cp, thal va ca, ba por tekrar tarin maghadir por mishe chon fasele beine maghadire in sotun ha maanaye riazi nadaran.

Group-based Imputation: maghadir gom shode ba estefade az mean va median har gorooh ke bar asase vizhegihaye dige taarif shodan jaigozin mishan. in ravesh deghate bishtari dare chon tafavot jamiati ro dar nazar migire.

## سوال 9

یکی از روش‌ها را انتخاب کرده و پیاده‌سازی کنید.

---

In [34]:
numeric_cols_with_na = df.select_dtypes(include=[np.number]).columns[df.select_dtypes(include=[np.number]).isnull().any()]
for col in numeric_cols_with_na:
    median_value = df[col].median()
    df[col].fillna(median_value, inplace=True)

chon maghadire null nadarim be karemoon nemiad. 

## سوال 10

رکوردهای تکراری را شناسایی کنید.

---

In [36]:
duplicate_count = df.duplicated().sum()
print("duplicated values: ", duplicate_count)

duplicated values:  1


## سوال 11

نسخه‌ای از داده بدون رکوردهای تکراری ایجاد کنید.

---

In [38]:
df_no_duplicates = df.drop_duplicates().reset_index(drop=True)

## سوال 12

با استفاده از Boxplot و IQR:

Outlier های هر ویژگی عددی را پیدا کنید.

---

In [41]:
# for col in df.select_dtypes(include=[np.number]):
#     Q1 = df[col].quantile(0.25)
#     Q3 = df[col].quantile(0.75)
#     IQR = Q3 - Q1
#     lower = Q1 - 1.5 * IQR
#     upper = Q3 + 1.5 * IQR
#     outliers = df[(df[col] < lower) | (df[col] > upper)]
#     print(f"{col}: {len(outliers)} outliers |  allowed range: [{lower:.1f}, {upper:.1f}]")

## سوال 13

جدولی ایجاد کنید:

| Feature | Number of Outliers |

---

## سوال 14

برای Outlier ها سه استراتژی مختلف پیشنهاد دهید.

---

## سوال 15

یک استراتژی را انتخاب و پیاده‌سازی کنید.

---

# بخش سوم

# Data Construction

---

در این بخش باید ویژگی‌های جدید ایجاد شوند.

---

## سوال 16

سن را به سه گروه تقسیم کنید:

* Young
* Middle Age
* Senior

ستون جدید:

```python
age_group
```

---

## سوال 17

برای chol سطح ریسک ایجاد کنید:

* Low
* Normal
* High

---

## سوال 18

برای trestbps سطح فشار خون تعریف کنید.

---

## سوال 19

حداقل دو Feature جدید طراحی کنید.

دلایل کسب‌وکاری آن را توضیح دهید.

---

## سوال 20

جدول زیر را تکمیل کنید:

| New Feature | Business Reason |

---

# بخش چهارم

# Data Integration

---

فرض کنید داده جدیدی از آزمایشگاه دریافت شده است.

فایل فرضی:

```python
lab_results.csv
```

---

## سوال 21

یک DataFrame آزمایشی بسازید.

---

## سوال 22

یک کلید مشترک تعریف کنید.

---

## سوال 23

Merge را انجام دهید.

---

## سوال 24

انواع Merge زیر را مقایسه کنید:

* inner
* left
* right
* outer

---

## سوال 25

مشخص کنید برای این پروژه کدام نوع Merge مناسب‌تر است.

---

# بخش پنجم

# Data Formatting

---

## سوال 26

نوع داده تمامی ستون‌ها را بررسی کنید.

---

## سوال 27

ستون‌هایی که باید Category باشند را شناسایی کنید.

---

## سوال 28

Datatype مناسب را اعمال کنید.

---

## سوال 29

حجم حافظه قبل و بعد از تغییر datatype را مقایسه کنید.

---

## سوال 30

ستون‌های متنی را استانداردسازی کنید.

مثال:

```python
Male
male
MALE
```

---

# بخش ششم

# Data Reduction

---

## سوال 31

ویژگی‌های دارای واریانس بسیار کم را شناسایی کنید.

---

## سوال 32

ویژگی‌های بسیار همبسته را شناسایی کنید.

---

## سوال 33

Heatmap جدید رسم کنید.

---

## سوال 34

اگر دو ویژگی همبستگی بالای 0.85 دارند:

کدام را حذف می‌کنید؟

چرا؟

---

## سوال 35

یک نسخه کاهش‌یافته از دیتاست ایجاد کنید.

نام:

```python
reduced_df
```

---

# بخش هفتم

# Data Transformation

---

# Normalization & Scaling

---

## سوال 36

بررسی کنید آیا ویژگی‌های عددی مقیاس‌های متفاوت دارند یا خیر.

---

## سوال 37

برای تمام ویژگی‌های عددی:

Min-Max Scaling را با NumPy پیاده‌سازی کنید.

فرمول:

```python
(x-min)/(max-min)
```

---

## سوال 38

برای تمام ویژگی‌های عددی:

Standardization را با NumPy پیاده‌سازی کنید.

---

## سوال 39

قبل و بعد از Scaling:

Histogram رسم کنید.

---

## سوال 40

مقایسه کنید:

* MinMax
* Standardization

برای این پروژه کدام مناسب‌تر است؟

---

# بخش هشتم

# Dataset Validation

---

## سوال 41

شکل نهایی دیتاست را گزارش کنید.

---

## سوال 42

آیا Null باقی مانده است؟

---

## سوال 43

آیا Duplicate باقی مانده است؟

---

## سوال 44

آیا Outlier ها مدیریت شده‌اند؟

---

## سوال 45

آیا تمامی ویژگی‌ها فرمت مناسبی دارند؟

---

## سوال 46

آیا داده برای KNN مناسب است؟

توضیح دهید.

---

## سوال 47

آیا داده برای Logistic Regression مناسب است؟

توضیح دهید.

---

## سوال 48

آیا داده برای Decision Tree مناسب است؟

توضیح دهید.

---

## سوال 49

دیتاست نهایی را ذخیره کنید:

```python
heart_prepared.csv
```

---

## سوال 50

یک گزارش مدیریتی تهیه کنید.

به سوالات زیر پاسخ دهید:

1. چه تغییراتی روی داده انجام شد؟
2. چه ویژگی‌هایی حذف شدند؟
3. چه ویژگی‌هایی ساخته شدند؟
4. چه مشکلاتی برطرف شد؟
5. آیا داده برای Modeling آماده است؟

---

# تحویل پروژه

فایل‌های زیر باید ارسال شوند:

```python
task002_data_preparation.ipynb
heart_prepared.csv
```

---

# هدف آموزشی این تسک

پس از پایان این تمرین دانشجو باید بتواند:

✅ Data Selection انجام دهد

✅ Data Cleaning انجام دهد

✅ Feature Construction انجام دهد

✅ Data Integration انجام دهد

✅ Data Formatting انجام دهد

✅ Data Reduction انجام دهد

✅ Scaling و Normalization انجام دهد

✅ یک دیتاست خام را به یک دیتاست آماده مدل‌سازی تبدیل کند

این دقیقاً خروجی مورد انتظار مرحله Data Preparation در فرآیند CRISP-DM است.